added AllKNN under-sampling method

dvro opened the pull request:
- Added AllKNN under-sampling method
    - imblearn/under_sampling/edited_nearest_neighbours.py. 
- changed added AllKNN in under_sampling/__init__.py for import purposes;
- added a plot example comparing ENN, RENN and AllKNN
    - examples/under-sampling/plot_allknn.py

`python plot_allknn.py`

> ENN
> Reduced 5.64\%
> RENN
> Reduced 8.36\%
> AllKNN
> Reduced 6.94\%

![allknn](https://cloud.githubusercontent.com/assets/511783/16847637/6f6b2736-49c8-11e6-8024-02471d53a019.png)

glemaitre commented:
@dvro Could you also write the test? You can check the other file to check how it is done for the moment.

glemaitre commented:
You should also avoid to merge and use rebase I think. The merge will be on our side.

dvro commented:
Ok, I'll do that today!

dvro commented:
@glemaitre I believe it is fine now; let me know if it still needs any improvement!

glemaitre commented:
Code wise it is fine. Just a last addition: can you add to the changelog in `doc/whats_new.rst`

dvro commented:
@glemaitre done

updated `README.rst`, `doc/whats_new.rst`, `doc/todo.rst`
changed to commit name to a simple `"AllKNN under-sampling technique"`

glemaitre commented:
I got an issue after merging on master with the testing on my repo. The table were mismatching.
Could you investigate. I put the inside a branch, you can [check](Travis https://travis-ci.org/glemaitre/imbalanced-learn/builds/145472835)

dvro commented:
That's weird ... everything seems to be fine on my end ... 

`Doctest: imblearn.pipeline.Pipeline ... ok
Doctest: imblearn.under_sampling.cluster_centroids.ClusterCentroids ... ok
Doctest: imblearn.under_sampling.condensed_nearest_neighbour.CondensedNearestNeighbour ... ok
Doctest: imblearn.under_sampling.edited_nearest_neighbours.AllKNN ... ok
Doctest: imblearn.under_sampling.edited_nearest_neighbours.EditedNearestNeighbours ... ok
Doctest: imblearn.under_sampling.edited_nearest_neighbours.RepeatedEditedNearestNeighbours ... ok
Doctest: imblearn.under_sampling.instance_hardness_threshold.InstanceHardnessThreshold ... ok
Doctest: imblearn.under_sampling.nearmiss.NearMiss ... ok
Doctest: imblearn.under_sampling.neighbourhood_cleaning_rule.NeighbourhoodCleaningRule ... ok
Doctest: imblearn.under_sampling.one_sided_selection.OneSidedSelection ... ok
Doctest: imblearn.under_sampling.random_under_sampler.RandomUnderSampler ... ok
Doctest: imblearn.under_sampling.tomek_links.TomekLinks ... ok
`

I'll look into it!

glemaitre commented on imblearn/under_sampling/tests/test_allknn.py:
```diff
@@ -0,0 +1,129 @@
+    currdir = os.path.dirname(os.path.abspath(__file__))
+    X_gt = np.load(os.path.join(currdir, 'data', 'allknn_x.npy'))
+    y_gt = np.load(os.path.join(currdir, 'data', 'allknn_y.npy'))
+    assert_array_equal(X_resampled, X_gt)
```
@dvro Try to use `assert_array_almost_equal` to see if this is not a problem of rounding.

dvro commented:
@glemaitre check out this weirdest thing:
On my env, the test_edited_nearest_neighbors fail:
`
(clean-env) dayvid:tests$ nosetests test_edited_nearest_neighbours.py
.....FFF.
======================================================================
FAIL: Test the fit sample routine
----------------------------------------------------------------------
`
and the test_allknn.py passes:

`
(clean-env) dayvid:tests$ nosetests test_allknn.py 
.........
----------------------------------------------------------------------
Ran 9 tests in 6.501s

OK
`

I created a clean-env using the same python version, numpy/scipy version, but the problem persists, specifically `Arrays are not equal (mismatch 2.82770406112%)`

do you think you could generate the `allknn_*.npy` files on your machine? and pull request them to my allknn repo?

Thanks,

glemaitre commented:
@dvro I am doing that.

glemaitre commented:
@dvro done

glemaitre commented:
ok so that's pretty weird, which OS are you using?

glemaitre commented on imblearn/under_sampling/edited_nearest_neighbours.py:
```diff
@@ -382,3 +382,171 @@ def _sample(self, X, y):
+    >>> from collections import Counter
+    >>> from sklearn.datasets import fetch_mldata
+    >>> from imblearn.under_sampling import AllKNN
+    >>> pima = fetch_mldata('diabetes_scale')
```
Can you change the example to avoid fetching data. It can be a problem at testing.

dvro replied:
done.

glemaitre commented on imblearn/under_sampling/edited_nearest_neighbours.py:
```diff
@@ -382,3 +382,171 @@ def _sample(self, X, y):
+            # updating ENN size_ngh
+            self.enn_.size_ngh = curr_size_ngh
+            if self.return_indices:
+                X_, y_, idx_ = self.enn_.fit_sample(X_, y_)
```
I am not sure that the indexes are correct. The indexes will be taken from `X_` iteratively which after several iteration are not corresponding to the indexes of `X`.

I think that at the first iteration `idx_` is correct then you need to pick up `idx_[idx_]`, so you will need a temporary variable.

glemaitre commented on imblearn/under_sampling/edited_nearest_neighbours.py:
```diff
@@ -382,3 +382,171 @@ def _sample(self, X, y):
+        if self.kind_sel not in SEL_KIND:
+            raise NotImplementedError
+
+        X_, y_ = X.copy(), y.copy()
```
Do we need a copy of `X` and `y` ?

dvro replied:
fixed.

dvro commented:
@glemaitre whats up with this appveyor, it seems like several tests are failing ...

glemaitre commented:
@dvro appveyor is still not working because of the npy file. That is why we have to move away from this solution.

This is fine for merging for me. Do you think that we need `X_ = X.copy()` in RENN as well?
